From 75062a325f284a787a3a504014c9bef8d8b3729f Mon Sep 17 00:00:00 2001
From: Petr Uzel <petr.uzel@suse.cz>
Date: Tue, 14 Jun 2016 13:17:00 +0200
Subject: Add support for NVMe devices

Recognize NVMe Devices, so "parted -s /dev/nvme0n1" now prints
"NVMe Device (nvme)" instead of "Model: Unknown (unknown)".

In order for a device to be recognized as NVMe, it has to
have a 'blkext' major number. But since this major can be
used also by other device types, we also check the device
path contains 'nvme' as a substring.

* NEWS: Mention the change
* include/parted/device.h.in(PedDeviceType): Add PED_DEVICE_NVME
* libparted/arch/linux.c(BLKEXT_MAJOR): New define.
* libparted/arch/linux.c(_is_blkext_major): New function.
* libparted/arch/linux.c(_device_probe_type): Recognize NVMe devices.
* libparted/arch/linux.c(linux_new): Handle NVMe devices.
* parted/parted.c(do_print): Add "nvme" to list of transports.

Signed-off-by: Brian C. Lane <bcl@redhat.com>

Origin: upstream, https://git.savannah.gnu.org/cgit/parted.git/commit?id=e4ae4330f3e33201aeeed3b7ca88e15d98d03e13
Bug-Debian: https://bugs.debian.org/886960
Bug-Ubuntu: https://bugs.launchpad.net/bugs/1742787
Last-Update: 2018-01-17

Patch-Name: nvme-devices.patch
---
 include/parted/device.h    |  3 ++-
 include/parted/device.in.h |  3 ++-
 libparted/arch/linux.c     | 14 ++++++++++++++
 parted/parted.c            |  2 +-
 4 files changed, 19 insertions(+), 3 deletions(-)

diff --git a/include/parted/device.h b/include/parted/device.h
index 798fb67a..aeaebb58 100644
--- a/include/parted/device.h
+++ b/include/parted/device.h
@@ -49,7 +49,8 @@ typedef enum {
         PED_DEVICE_VIRTBLK      = 15,
         PED_DEVICE_AOE          = 16,
         PED_DEVICE_MD           = 17,
-        PED_DEVICE_LOOP         = 18
+        PED_DEVICE_LOOP         = 18,
+        PED_DEVICE_NVME         = 19
 } PedDeviceType;
 
 typedef struct _PedDevice PedDevice;
diff --git a/include/parted/device.in.h b/include/parted/device.in.h
index 82d4104e..d38db44a 100644
--- a/include/parted/device.in.h
+++ b/include/parted/device.in.h
@@ -49,7 +49,8 @@ typedef enum {
         PED_DEVICE_VIRTBLK      = 15,
         PED_DEVICE_AOE          = 16,
         PED_DEVICE_MD           = 17,
-        PED_DEVICE_LOOP         = 18
+        PED_DEVICE_LOOP         = 18,
+        PED_DEVICE_NVME         = 19
 } PedDeviceType;
 
 typedef struct _PedDevice PedDevice;
diff --git a/libparted/arch/linux.c b/libparted/arch/linux.c
index d91ad9a0..9b319de7 100644
--- a/libparted/arch/linux.c
+++ b/libparted/arch/linux.c
@@ -279,6 +279,7 @@ struct blkdev_ioctl_param {
 #define SDMMC_MAJOR             179
 #define LOOP_MAJOR              7
 #define MD_MAJOR                9
+#define BLKEXT_MAJOR            259
 
 #define SCSI_BLK_MAJOR(M) (                                             \
                 (M) == SCSI_DISK0_MAJOR                                 \
@@ -440,6 +441,12 @@ _is_virtblk_major (int major)
         return _major_type_in_devices (major, "virtblk");
 }
 
+static int
+_is_blkext_major (int major)
+{
+        return _major_type_in_devices (major, "blkext");
+}
+
 #ifdef ENABLE_DEVICE_MAPPER
 static int
 _dm_task_run_wait (struct dm_task *task, uint32_t cookie)
@@ -691,6 +698,8 @@ _device_probe_type (PedDevice* dev)
                 dev->type = PED_DEVICE_LOOP;
         } else if (dev_major == MD_MAJOR) {
                 dev->type = PED_DEVICE_MD;
+        } else if (_is_blkext_major(dev_major) && dev->path && strstr(dev->path, "nvme")) {
+                dev->type = PED_DEVICE_NVME;
         } else {
                 dev->type = PED_DEVICE_UNKNOWN;
         }
@@ -1454,6 +1463,11 @@ linux_new (const char* path)
                         goto error_free_arch_specific;
                 break;
 
+        case PED_DEVICE_NVME:
+                if (!init_generic (dev, _("NVMe Device")))
+                        goto error_free_arch_specific;
+                break;
+
         case PED_DEVICE_ATARAID:
                 if (!init_generic (dev, _("ATARAID Controller")))
                         goto error_free_arch_specific;
diff --git a/parted/parted.c b/parted/parted.c
index f27a0358..e8532d8b 100644
--- a/parted/parted.c
+++ b/parted/parted.c
@@ -972,7 +972,7 @@ _print_disk_info (const PedDevice *dev, const PedDisk *diskp)
                                          "cpqarray", "file", "ataraid", "i2o",
                                          "ubd", "dasd", "viodasd", "sx8", "dm",
                                          "xvd", "sd/mmc", "virtblk", "aoe",
-                                         "md", "loopback"};
+                                         "md", "loopback", "nvme"};
 
         char* start = ped_unit_format (dev, 0);
         PedUnit default_unit = ped_unit_get_default ();
